src/S3/S3Transfer/S3TransferManager.php

```diff
@@ -0,0 +1,1048 @@
+        ?TransferListener $progressTracker = null,
+    ): PromiseInterface
+    {
+        if (!file_exists($destinationDirectory)) {
+            throw new InvalidArgumentException(
+                "Destination directory `$destinationDirectory` MUST exists."
+            );
+        }
+
+        $bucket = $this->parseBucket($bucket);
+
+        if ($progressTracker === null
+            && ($config['track_progress'] ?? $this->config['track_progress'])) {
+            $progressTracker = new MultiProgressTracker();
+        }
+
+        $listArgs = [
+            'Bucket' => $bucket
+        ]  + ($config['list_object_v2_args'] ?? []);
+        if (isset($config['s3_prefix']) && !isset($listArgs['Prefix'])) {
+            $listArgs['Prefix'] = $config['s3_prefix'];
+        }
+
+        if (isset($config['s3_delimiter']) && !isset($listArgs['Delimiter'])) {
+            $listArgs['Delimiter'] = $config['s3_delimiter'];
+        }
+
+        $objects = $this->s3Client
+            ->getPaginator('ListObjectsV2', $listArgs)
+            ->search('Contents[].Key');
+        if (isset($config['filter'])) {
+            if (!is_callable($config['filter'])) {
+                throw new InvalidArgumentException("The parameter \$config['filter'] must be callable.");
+            }
+
+            $filter = $config['filter'];
+            $objects = filter($objects, function (string $key) use ($filter) {
+                return call_user_func($filter, $key);
+            });
+        }
+
+        $objects = map($objects, function (string $key) use ($bucket) {
+            return  "s3://$bucket/$key";
+        });
+        $getObjectRequestCallback = null;
+        if (isset($config['get_object_request_callback'])) {
+            if (!is_callable($config['get_object_request_callback'])) {
+                throw new InvalidArgumentException(
+                    "The parameter \$config['get_object_request_callback'] must be callable."
+                );
+            }
+
+            $getObjectRequestCallback = $config['get_object_request_callback'];
+        }
+
+        $failurePolicyCallback = null;
+        if (isset($config['failure_policy']) && !is_callable($config['failure_policy'])) {
+            throw new InvalidArgumentException(
+                "The parameter \$config['failure_policy'] must be callable."
+            );
+        } elseif (isset($config['failure_policy'])) {
+            $failurePolicyCallback = $config['failure_policy'];
+        }
+
+        $promises = [];
+        $objectsDownloaded = 0;
+        $objectsFailed = 0;
+        foreach ($objects as $object) {
+            $objectKey = $this->s3UriAsBucketAndKey($object)['Key'];
+            $destinationFile = $destinationDirectory . '/' . $objectKey;
+            if ($this->resolvesOutsideTargetDirectory($destinationFile, $objectKey)) {
+                throw new S3TransferException(
+                    "Cannot download key ' . $objectKey
```

Comment:
We should have some comprehensive tests around this considering the security implications

Reply:
Done.